src/projection.jl

```diff
@@ -52,93 +128,197 @@ function ProjectTo(x::T) where {T<:Union{<:Tuple,NamedTuple}}
 end
 
 # Generic
-(project::ProjectTo)(dx::AbstractThunk) = project(unthunk(dx))
 (::ProjectTo{T})(dx::T) where {T} = dx  # not always true, but we can special case for when it isn't
-(::ProjectTo{T})(dx::AbstractZero) where {T} = zero(T)
+(::ProjectTo{T})(dx::AbstractZero) where {T} = dx # zero(T)
+
+# Thunks
+(project::ProjectTo)(dx::Thunk) = Thunk(project ∘ dx.f)
+(project::ProjectTo)(dx::InplaceableThunk) = Thunk(project ∘ dx.val.f) # can't update in-place part
+(project::ProjectTo)(dx::AbstractThunk) = project(unthunk(dx))
 
-# Number
-ProjectTo(::T) where {T<:Number} = ProjectTo{T}()
+# Non-differentiable -- not so sure this is necessary. Keeping value like this is a bit awkward.
+for T in (:Bool, :Symbol, :Char, :String, :Val, :Type)
+    @eval ProjectTo(dx::$T) = ProjectTo{AbstractZero}(; value=dx)
+end
+(::ProjectTo{AbstractZero})(dx::AbstractZero) = dx
+(::ProjectTo{AbstractZero})(dx) = ZeroTangent()
+
+# Number -- presently preserves Float32, but not Int, is this ideal?
+ProjectTo(::T) where {T<:Number} = ProjectTo{float(T)}()
 (::ProjectTo{T})(dx::Number) where {T<:Number} = convert(T, dx)
 (::ProjectTo{T})(dx::Number) where {T<:Real} = convert(T, real(dx))
 
+ProjectTo(::Type{T}) where T<:Number = ProjectTo(zero(T)) # maybe
+
 # Arrays 
-ProjectTo(xs::T) where {T<:Array} = ProjectTo{T}(; elements=map(ProjectTo, xs))
-function (project::ProjectTo{T})(dx::Array) where {T<:Array}
-    _call(f, x) = f(x)
-    return T(map(_call, project.elements, dx))
-end
-function (project::ProjectTo{T})(dx::AbstractZero) where {T<:Array}
-    return T(map(proj -> proj(dx), project.elements))
-end
-(project::ProjectTo{<:Array})(dx::AbstractArray) = project(collect(dx))
+# ProjectTo(xs::T) where {T<:Array} = ProjectTo{T}(; elements=map(ProjectTo, xs))
+# function (project::ProjectTo{T})(dx::Array) where {T<:Array}
+#     _call(f, x) = f(x)
+#     return T(map(_call, project.elements, dx))
+# end
+# function (project::ProjectTo{T})(dx::AbstractZero) where {T<:Array}
+#     return T(map(proj -> proj(dx), project.elements))
+# end
+# (project::ProjectTo{<:Array})(dx::AbstractArray) = project(collect(dx))
 
 # Arrays{<:Number}: optimized case so we don't need a projector per element
-function ProjectTo(x::T) where {E<:Number,T<:Array{E}}
-    return ProjectTo{T}(; element=ProjectTo(zero(E)), size=size(x))
+function ProjectTo(x::AbstractArray{T,N}) where {T<:Number,N}
+    T == Bool && return ProjectTo(false)
+    sub = ProjectTo(zero(T))
+    return ProjectTo{AbstractArray{eltype(sub),N}}(; element=sub, axes=axes(x))
 end
-(project::ProjectTo{<:Array{T}})(dx::Array) where {T<:Number} = project.element.(dx)
-function (project::ProjectTo{<:Array{T}})(dx::AbstractZero) where {T<:Number}
-    return zeros(T, project.size)
+function (project::ProjectTo{AbstractArray{T,N}})(dx::AbstractArray{S,M}) where {T,S,N,M}
```

Comment:
Quite possibly. Right now `ProjectTo{AbstractArray` can only be constructed for numbers. Arrays of other things I kicked down the road for now.

Reply:
The code we have right now in `#master` does support arrays of arrays.
Arrays of arrays do exist in the wild, I keep having to do things about that to avoid breaking Zygote.

Reply:
Yes, I agree. I just didn't try yet in this PR.